### PR TITLE
introduce flaky lines, remove unnecessary functionality

### PR DESCRIPTION
**Pull request by nikitaevg**
We introduce flaky indicators that help distinguish flakes from real failures. We also remove all functionality that is not necessary for our project. And we also add test coverage for the action.

### CODE REVIEW

File: README.md
Comment (seanlip) [inline README.md]: Do these have to be full lines or just substrings?

File: src/action.spec.ts
Comment (seanlip) [inline src/action.spec.ts:19]: Add copyright notice.

File: src/index.ts
Comment (seanlip) [inline src/index.ts]: Shouldn't the exit code be 1?

File: src/action.spec.ts
Comment (seanlip) [inline src/action.spec.ts:122]: I am getting a bit confused when reading these tests. In some cases the flaky string seems to be inside the temp files and in others it seems to be within the command / printed to the terminal. So I don't entirely understand what is going on here...

If the point of the second file is to print stuff to the terminal output then why not just echo it like we do with everything else?

File: src/action.spec.ts
Comment (seanlip) [inline src/action.spec.ts:48]: Might be worth adding a test to distinguish matching by substring vs matching by line

File: src/action.ts
Comment (seanlip) [inline src/action.ts]: Should this be 1?

File: src/action.ts
Comment (seanlip) [inline src/action.ts:91]: It does seem like this is checking for substrings, maybe we should be calling the argument flaky_test_indicators or flaky_test_indicator_substrings instead (or something that doesn't have "lines" in the name).

File: src/action.ts
Comment (seanlip) [inline src/action.ts:117]: Maybe add log after this to say it failed flakily and that we're restarting.

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @nikitaevg, sorry for the review delay! This looks good on the whole IMO :)

I just had a few comments -- haven't had time to make changes directly yet, but will try to make a PR later in the week if possible, so I figured I'd just leave the comments here for reference at least.

File: README.md
Comment (nikitaevg) [reply README.md]: Just substrings, updated the description

File: src/action.spec.ts
Comment (nikitaevg) [reply src/action.spec.ts:48]: Changed this test so it shows that we look for substrings

File: src/action.spec.ts
Comment (nikitaevg) [reply src/action.spec.ts:122]: > In some cases the flaky string seems to be inside the temp files and in others it seems to be within the command / printed to the terminal.

It's always printed to the terminal, see this tests `cat`s the file contents.

> If the point of the second file is to print stuff to the terminal output then why not just echo it like we do with everything else?

We need different outputs in the first and second attempts - the first attempt is a flake, the second is not. Adjusted the comments a bit.

File: src/action.ts
Comment (nikitaevg) [reply src/action.ts]: Yeah, you're right

File: src/action.ts
Comment (nikitaevg) [reply src/action.ts:91]: Renamed to substrings_indicating_flaky_execution

File: src/action.ts
Comment (nikitaevg) [reply src/action.ts:117]: I don't mind, but note that we already will have
"Output contains flaky line: ${flakyIndicator}" log + "Starting attempt #${attempt}". IMO it's enough, but I can add if you want

File: src/index.ts
Comment (nikitaevg) [reply src/index.ts]: Yep, thanks!

File: src/action.spec.ts
Comment (nikitaevg) [reply src/action.spec.ts:19]: Done

**Review by nikitaevg — COMMENTED** [review]
@seanlip PTAL

File: src/action.ts
Comment (seanlip) [reply src/action.ts:117]: Ah thanks, I missed that, sorry -- but I think the call to info() is in the wrong place. Can we move it to this function?

Based on its name, hasFlakyOutput() is something that would be expected to just return a boolean and shouldn't have any meaningful side effects, whereas we do want printing the log to be a deliberate action that we take in the execution.

I also suggest adding "Restarting test" or similar to that message since it connects the restart with the error.

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @nikitaevg -- just one more thing and then I think it's good to go!

Also please update the first comment of this PR thread with a description of changes, thanks.

File: src/action.ts
Comment (nikitaevg) [reply src/action.ts:117]: > Based on its name, hasFlakyOutput() is something that would be expected to just return a boolean and shouldn't have any meaningful side effects, whereas we do want printing the log to be a deliberate action that we take in the execution.

I agree with that, but I think it's very helpful to output which flake indicator was found, so we need to output something in the hasFlakyOutput method.

I moved logging to the calling side, and also left the log of what indicator was found.

**Comment by nikitaevg** [comment]
@seanlip PTAL!

**Review by seanlip — COMMENTED** [review]
Hi @nikitaevg -- this looks good, thanks! But I want to get the tests to run. Do you know how to add tests to the CI so that we can ensure this (and subsequent changes don't break anything)?

I'll try closing and reopening this PR to see if that works...

**Comment by seanlip** [comment]
Ok, hm. Could you go to .github/workflows/ci_cd.yml and clean up the unnecessary stuff?  You can view runs here: https://github.com/oppia/retry/actions

**Comment by nikitaevg** [comment]
> Ok, hm. Could you go to .github/workflows/ci_cd.yml and clean up the unnecessary stuff? You can view runs here: https://github.com/oppia/retry/actions

Ohhh, I didn't notice that file. It has so many tests, but I think the tests I provided are good enough.

So I want to leave `ci_unit` only and remove other jobs, would that be ok for you?

Also, maybe I'll do it in a separate PR? This PR becomes too big.

**Comment by nikitaevg** [comment]
@seanlip PTAL

**Comment by seanlip** [comment]
I suggest keeping ci_unit but dropping the codecov part. I defer to your judgment on integration -- but in general I suggest keeping tests that align with the remaining functionality (ok to write new ones in a separate PR).

Agree in general with not wanting PRs to be too big, but I think we do need some tests to run on presubmit before we can merge this. So I suggest fixing that in this PR since this is where you also modify the behaviour (I thought of doing them in a separate PR that gets merged before this one but I think that wouldn't work).

**Comment by nikitaevg** [comment]
@seanlip Let me remove all integration tests in this PR and add them in the following one, sounds good? I fixed the github actions, see the latest commit.

**Review by seanlip — APPROVED** [review]
Yup seems good. Thanks!

**Pull request merged**
